doc/developer/design/20210714_reclocking.md

```diff
@@ -0,0 +1,98 @@
+# Reclocking
+
+## Summary
+
+"Reclocking" a stream of data is (for the purposes of this document) translating its events from one time domain to another time domain.
+
+Many of our sources of data come with their own understanding of forward progress.
+Often, these are incompatible with each other, and with native notions of time within Materialize.
+For example,
+    Kafka topics contain multiple "parts", each of which advances through "offsets".
+    Postgres sources come from a log where entries have a "log sequence number" (LSN).
+    File sources have a line number.
+    Materialize CDC sources have explicit timestamps.
+
+The source-specific "timestamps" of forward progress are useful, but they are not directly comparable.
+This document describes one framework for converting between the timestamps including in to and potentially out of Materialize's system time.
+
+## Goals
+
+The primary goal of this document is to describe a formal way to translate collections that evolve with one gauge into collections that evolve with another gauge.
+That formalism should have the properties that:
+1. There is an explicit representation of the translation (rather than a behavioral description).
+2. The translation should be sidecar metadata, rather than a rewriting of the source data.
+3. The combination of durable source data and durable metadata should result in a durable reclocked stream.
+4. The sidecar metadata should be compactable whenever the reclocked stream is compactable.
+
+The formalism should support the goals of
+A. Permitting the use of streams from one timeline in another timeline.
+B. Describing durable state that can make sources like Kafka, Postgres, and files NONVOLATILE in the strong sense that their contents can be replayed exactly.
```

Comment:
I can add `in a meaningful timeline` between `replayed` and `exactly`. I don't think there is confusion here that e.g. Kafka's `(partition, offset)` is not actually a timestamp for a timeline that we can use in Materialize. I'll rewrite as `gauge` rather than `timestamp` to try and avoid the confusion.

Reply:
Changed to "in Materialize's system timeline" as this list is only meant to be representative.